src/weights.jl

```diff
@@ -291,3 +291,96 @@ end
 
 wmedian(v::RealVector, w::RealVector) = median(v, weights(w))
 wmedian{W<:Real}(v::RealVector, w::WeightVec{W}) = median(v, w)
+
+
+
+###### Weighted quantile #####
+# Variables m, g, low, high, correspond respectively to m, gamma, x[j], x[j+1] in the R help file for quantile
+
+function Base.quantile{W<:Real}(v::RealVector, p::RealVector, w::WeightVec{W})
+
+    isempty(v) && error("quantile of an empty array is undefined")
+    isempty(p) && throw(ArgumentError("empty quantile array"))
+
+    ppermute = sortperm(p)
+    p = p[ppermute]
+
+    # make sure the quantiles are in [0,1]
+    p = bound_quantiles(p)
+
+    if w.sum == 0
+        error("weight vector cannot sum to zero")
+    end
+    if length(v) != length(w)
+        error("data and weight vectors must be the same size. got $(length(v)) and $(length(w))")
+    end
+    for x in w.values
+        isnan(x) && error("weight vector cannot contain NaN entries")
+        x < 0 && error("weight vector cannot contain negative entries")
+    end
+    for x in v
+        isnan(x) && return x
+    end
+    lv = length(v)
+    lp = length(p)
+    m = 1 - p
+    index = (lv * p + m) * w.sum / lv
+    wt = w.values
+    vpermute = sortperm(v)
+    low = fill(v[vpermute[end]], lp)
+    high = fill(v[vpermute[end]], lp)
+    g = fill(0.0, lp)
+    cumulative_weight = zero(eltype(wt))
+    k = 1
+    i = 0
+    while (k <= lp) & (i <= lv)
+        if cumulative_weight > index[k]
+            g[k] = (index[k] - (cumulative_weight - wt[vpermute[i]])) * lv / w.sum
+            high[k] = v[vpermute[i]]
+            low[k] = v[vpermute[max(i-1, 1)]]  
+            # Below we catch the case where an element equal to v[vpermute[i]] has a weight lower than threshold. 
+            # For instance median in (1, 2, 2) (0.1, 0.8, 0.1)
+            # In this case, we just need to set low[k] = high[k]
+            if low[k] != high[k] 
+                l = 1
+                wthreshold = cumulative_weight - wt[vpermute[i-1]] - index[k] 
+                # loop over posterior elements with same value
+                while (i + l <= lv && v[vpermute[i+l]] == high[k])
+                    if wt[vpermute[i+l]] <= wthreshold
+                        low[k] = high[k]
+                        break
+                    end
+                    l += 1
+                end
+            end
+            k += 1
+        else
+            i += 1
+            if i <= lv
+                cumulative_weight += wt[vpermute[i]]
+            end
+        end
+    end
+    out = Array(Real, lp)
```

Comment:
Yes. In Julia, functions must generally be type-stable for performance. This means they should always return the same type given their input types. Since there's a possibility that we may need to return a float, we must always return this type.